Refactor mlk_polymat_permute_bitrev_to_custom

mkannwischer opened the pull request:
- Resolves #1375

This commit refactors mlk_polymat_permute_bitrev_to_custom to not require the helper function mlk_polyvec_permute_bitrev_to_custom. The function was only needed due CBMC limitations.

hanno-becker commented:
@mkannwischer Can you comment on state/plans for this?

willieyz commented:
> > @mkannwischer Can you comment on state/plans for this?
> 
> 
> 
> See https://github.com/pq-code-package/mldsa-native/pull/770 - we are waiting for https://github.com/diffblue/cbmc/pull/8705 to be merged.
> 
> 
> 
> @willieyz, could you rebase this PR and also include the experimental branch of CBMC so we can confirm that the proofs pass with that version. 
> 
> 
> 
> 
> 
> 

Yes, I can do it.

willieyz committed:
Refactor mlk_polymat_permute_bitrev_to_custom

This commit refactors mlk_polymat_permute_bitrev_to_custom to not require the
helper function mlk_polyvec_permute_bitrev_to_custom.
The function was only needed due CBMC limitations.

The code structure now mimics the data structure to make proof tractable.
Also updates Makefile for this proof in line with the similar function
in mldsa-native.

Signed-off-by: Matthias J. Kannwischer <matthias@kannwischer.eu>
Signed-off-by: Rod Chapman <rodchap@amazon.com>
Signed-off-by: willieyz <willie.zhao@chelpis.com>

Re-format with clang-format and add decreases contract

Signed-off-by: Rod Chapman <rodchap@amazon.com>

hanno-becker commented:
@mkannwischer I rebased this. Can you check that you're happy once CI is?
